### DISCUSSION TITLE
Label Documentation

### DISCUSSION THREAD

**Discussion by dpardi02** [Project-MONAI/MONAILabel#1582] [Q&A] [answered]
Is there a place I could look to find documentation regarding how to format my premade labels into MONAI Label? I have had struggles thus far where the labels I submit are not recognized and are not contributing to the training and learning. They are all Nifti files but there might be other parameters I have configured differently for my files. Thanks for the help.

**Comment by diazandr3s** [comment]
Hi @dpardi02,

Thanks for asking this.

In this video, I commented on how the volumes and labels should be organized to use MONAI Label: https://youtu.be/-HAryYAO5J4?si=OaZMhMQW9lh-Njvb&t=778

Here is more information: https://github.com/Project-MONAI/MONAILabel#local-datastore-for-the-radiology-app-on-single-modality-images

Are you working with non-overlapping segments? How many annotated volumes you've premade?

Hope this helps,

**Reply by dpardi02** [reply]
Hi @diazandr3s,

Thank you for your response. I am aware of how to organize the labels regarding which folder to put them in for them to be recognized. However, when I submit the labels I have made nothing indicates they are being recognized. I have not worked with datasets that are very large thus far, only about a total of 20 volumes where 2-3 have labels. The segments are non-overlapping and are binary.

**Reply by diazandr3s** [reply]
Hi @dpardi02,

Thanks for the update.
Sorry, what do you mean by `"when I submit the labels I have made nothing indicates they are being recognized"`? 

When you submit a label, it goes to the `labels/original` folder. Can you check this folder after submitting the label?

If this is not the issue, I'll need more information to understand what error or issue you are facing here.

**Reply by dpardi02** [reply]
Hi @diazandr3s 
Sure thing,
Part of my confusion may stem from an incomplete understanding of the process. I have been working with datasets of around 20. When I take the time to make labels within MONAI Label and 3DSlicer, the training and deep learning work fine. The progress and accuracy bars move accordingly and I have been able to get rough auto-segmented images from this process. 

However, I have many labels made from another source that I would like to apply to MONAI Label so that I can use its learning and training functions for future samples. When I go to put these in the labels/original folder and run the same training, there is no progression and the process ends almost instantly. 

I am wondering whether there is a format discrepancy between the types of labels being made by the built-in labeler and the kinds I am providing and whether there is a place I could go to see the accepted format and specifications for my labels.

I am more than happy to answer any additional questions or provide information about my data if needed.

Thanks,
Daniel

**Reply by diazandr3s** [reply]
Hi @dpardi02,

Thanks for clarifying.

> However, I have many labels made from another source that I would like to apply to MONAI Label so that I can use its learning and training functions for future samples. When I go to put these in the labels/original folder and run the same training, there is no progression and the process ends almost instantly.
 
Did you make sure the labels have exactly the same name as the volumes/images? Are the labels in the same format as the volumes?

Sorry, which process ends immediately? The training process?

> I am wondering whether there is a format discrepancy between the types of labels being made by the built-in labeler and the kinds I am providing and whether there is a place I could go to see the accepted format and specifications for my labels.

If both labels and volumes are in NIfTI/NRRD and have the same name, MONAI Label will recognise them without issues.

After copying the labels to labels/original, you should be able to see in the logs that those labels are being added. Could you please check that?

Let me know

**Reply by dpardi02** [reply]
Hi @diazandr3s,

The labels all have the exact same name as the volumes and are all the same format so I do not think that is the issue. 

Yes, the training process is the one ending immediately but only when the labels are being imported. To make sure it wasn't another issue I had previously taken the time to make a rough label inside Slicer and the training went on then. 

It is good to know that MONAI Label should be recognizing the NIfTIs so thank you for that information.

The labels do seem to be showing up in the logs so something is being read within the labels folder. I will continue to investigate but in the meantime, I have a few questions if you don't mind:

1. Exactly what is the labels/original folder used for? In what context would labels be put in that folder?
2. A somewhat unrelated question to the current issue but still along the lines of using MONAI Label for this segmentation. In some of the videos and demonstrations I have seen made by either you or other members of the MONAI team, the pre-trained model is set to 'FALSE' when starting a new from-scratch session. I understand why this is, but is there a pre-trained model that could be used without providing any labels that are simply built-in with MONAI? My project is currently investigating the use of MONAI Label from scratch where we upload our own previously made labels but if there is a model that would possibly speed this process along I think I would be interested in investigating it.

As always, thanks so much for your continued help.

Best,
Daniel

**Reply by diazandr3s** [reply]
Hi @dpardi02,

Yes, please double-check and let me know. It shouldn't be an error when adding more labels.

Regarding these questions:

> Exactly what is the labels/original folder used for? In what context would labels be put in that folder?

Very good question. In labels/original we store the predictions obtained by the model when running batch inference. Those predictions are still not approved or reviewed by the expert.
You could make the model to run inference and keep the predictions there. Then when you fetch the next sample, it'll show the volume and prediction at the same time. The idea is that the expert will review the predictions, and correct or approve them. Once the prediction is approved, it is moved to the labels/final folder.

> A somewhat unrelated question to the current issue but still along the lines of using MONAI Label for this segmentation. In some of the videos and demonstrations I have seen made by either you or other members of the MONAI team, the pre-trained model is set to 'FALSE' when starting a new from-scratch session. I understand why this is, but is there a pre-trained model that could be used without providing any labels that are simply built-in with MONAI? My project is currently investigating the use of MONAI Label from scratch where we upload our own previously made labels but if there is a model that would possibly speed this process along I think I would be interested in investigating it.

Another great question. Currently, the labels have to be defined before starting to use the model. 
However, having a foundational model that automatically adjusts/configures the regions the user wants to segment is something of interest. We are always happy to listen to ideas on how to improve usability with clinical motivations in mind.

If you think a video call helps to further discuss this, please let me know.

**Reply by dpardi02** [reply]
Hi @diazandr3s,

Thanks for that additional information. I am pleased to say we have figured out our label discrepancy and were able to correctly configure our labels to run the training. Thanks so much for all your help through this process.

Best,
Daniel

**Reply by diazandr3s** [reply]
Thanks for the update, @dpardi02. 
If you can, please share more details so others can benefit from this.

**Reply by dpardi02** [reply]
Sure thing @diazandr3s 

One issue was the orientation of some of our labels where we were exporting them from ended up switching planes in some instances so they had to be permuted back to match the raw data.

However, the main issue was the content of our labels. Upon further inspection, we found that some errors that were being given was pertaining to the use of multichannel labels. After converting our labels to strict binary we found success. 

Again thanks for your help. I have just one final question before I close this ticket if you don't mind. What customizable options are there to the training and does it all take place in the configs document? I know that the number of max epochs can be changed from here but was wondering if there were any other options. Additionally, say I open my dataset, run training, and then close the slicer and MONAI Label session, is there a way to see the saved training somewhere and apply it to images elsewhere? In other words, can the product of a training session be used at a different time or does it have to be applied to images then and there?

**Reply by diazandr3s** [reply, marked as answer]
Many thanks for the update @dpardi02
Glad you found the source of the problem.

About your questions:

> What customizable options are there to the training and does it all take place in the configs document? 

Yes, most of them are in the config file. However, you can go deeper and modify in the training file the learning rate, optimizer, loss function, pre-transforms (data augmentation), etc: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/lib/trainers/segmentation.py#L61

> is there a way to see the saved training somewhere and apply it to images elsewhere? In other words, can the product of a training session be used at a different time or does it have to be applied to images then and there?

When you trigger the training process, you'll find the best model in .pt format, several checkpoints and Tensorboard file that you can use to monitor the training - all this inside the folder **radiology/model**.

You can use the .pt file and run it as any PyTorch model. No need to specifically use MONAI Label.

Another way is by running MONAI Label without starting the server. For this, you can run inference using the main function of the Radiology app: https://github.com/Project-MONAI/MONAILabel/blob/main/sample-apps/radiology/main.py#L267

I hope this helps,

**Reply by dpardi02** [reply]
Perfect, thank you @diazandr3s for all the help with my questions. This has greatly progressed the work on my project. Much appreciated.
